bugsnag-android-core/src/main/java/com/bugsnag/android/RemoteConfig.kt

```diff
@@ -0,0 +1,109 @@
+package com.bugsnag.android
+
+import android.util.JsonReader
+import com.bugsnag.android.internal.DateUtils
+import java.util.Date
+
+internal class RemoteConfig(
+    val configurationTag: String,
+    val configurationExpiry: Date,
+    val discardRules: List<DiscardRule>
+) : JsonStream.Streamable {
+
+    override fun toStream(stream: JsonStream) {
+        stream.beginObject()
+        stream.name(KEY_CONFIGURATION_TAG).value(configurationTag)
+        stream.name(KEY_CONFIG_EXPIRY).value(DateUtils.toIso8601(configurationExpiry))
+        stream.name(KEY_DISCARD_RULES).beginArray()
+        for (rule in discardRules) {
+            stream.beginObject()
+            when (rule) {
+                is DiscardRule.All -> {
+                    stream.name(MATCH_TYPE).value(DISCARD_RULE_ALL)
+                }
+
+                is DiscardRule.AllHandled -> {
+                    stream.name(MATCH_TYPE).value(DISCARD_RULE_ALL_HANDLED)
+                }
+            }
+            stream.endObject()
+        }
+        stream.endArray()
+        stream.endObject()
+    }
+
+    internal companion object : JsonReadable<RemoteConfig?> {
+        internal const val KEY_CONFIGURATION_TAG = "configurationTag"
+        internal const val KEY_CONFIG_EXPIRY = "configurationExpiry"
+        internal const val KEY_DISCARD_RULES = "discardRules"
+        internal const val DISCARD_RULE_ALL = "ALL"
+        internal const val DISCARD_RULE_ALL_HANDLED = "ALL_HANDLED"
+        internal const val MATCH_TYPE = "match_type"
+
+        override fun fromReader(reader: JsonReader): RemoteConfig? {
+            var remoteConfig: RemoteConfig? = null
+            with(reader) {
+                beginObject()
+                var configurationTag: String? = null
+                var configExpiry: Date? = null
+                var discardRules: List<DiscardRule> = emptyList()
+
+                while (hasNext()) {
+                    val key = nextName()
+                    when (key) {
+                        KEY_CONFIGURATION_TAG -> configurationTag = nextString()
+                        KEY_CONFIG_EXPIRY -> configExpiry = DateUtils.fromIso8601(nextString())
+                        KEY_DISCARD_RULES -> {
+                            discardRules = parseDiscardRules(reader)
+                        }
+                    }
+                }
+                if (configurationTag != null && configExpiry != null) {
+                    remoteConfig = RemoteConfig(configurationTag, configExpiry, discardRules)
+                }
+                endObject()
+            }
+            return remoteConfig
+        }
+
+        private fun parseDiscardRules(reader: JsonReader): List<DiscardRule> {
```

Comment:
This function (as suggested by the linter rules) is too deeply nested. Making it an extension function of `JsonReader` will eliminate the `with(reader)` level, and giving `DiscardRule` a `JsonReadable<DiscardRule?>` companion will let us simplify this function.